worlds/mm2/rules.py

```diff
@@ -70,8 +70,16 @@
 }
 
 
-def can_defeat_enough_rbms(state: "CollectionState", player: int, required: int) -> bool:
-    return state.has_from_list(robot_masters.values(), player, required)
+def can_defeat_enough_rbms(state: "CollectionState", player: int,
+                           required: int, boss_requirements: Dict[int, List[str]]):
+    can_defeat = 0
+    for boss, reqs in boss_requirements.items():
+        if boss in robot_masters:
+            if state.has_all(map(lambda x: weapons_to_name[x], reqs), player):
+                can_defeat += 1
+                if can_defeat >= required:
+                    return True
+    return False
```

Comment:
Ftr, this game gens so fast so I'm definitely not gonna force this, but I think it'd be really nice if this could be further "pre-reduced". Consider this:

```
robot_master_reqs = {
    boss: reqs for boss, reqs in boss_requirements.items() if boss in robot_masters
}
world.robot_master_to_weapon_names = {
    map(lambda x: weapons_to_name[x], reqs) for boss, reqs in robot_master_reqs
}
```

```
def can_defeat_enough_rbms(state: "CollectionState", player: int,
                           required: int, robot_master_to_weapon_names: Dict[int, List[str]]):
    can_defeat = 0
    for boss, weapon_names in robot_master_to_weapon_names.items():
        if state.has_all(weapon_names, player):
            can_defeat += 1
            if can_defeat >= required:
                return True
    return False
```

Reply:
In fact, at that point, you don't even need the names of the robot masters anymore, do you! The variable `boss` is now unused in that for loop.

You can reduce it down to a **list** of only the weapon names lists that beat each robot master somewhere, but omit the information about *which* robot master they belong to bc that's not relevant to the logic:

```
world.robot_master_weapon_lists = [
    map(lambda x: weapons_to_name[x], weapons_that_kill_it)
    for boss, weapons_that_kill_it in boss_requirements.items() 
    if boss in robot_masters
]
```

and then your rbm function can be:

```
def can_defeat_enough_rbms(state: "CollectionState", player: int,
                           required: int, robot_master_weapon_lists: List[List[str]]):
    can_defeat = 0
    for weapon_list in robot_master_weapon_lists:
        if state.has_all(weapon_list, player):
            can_defeat += 1
            if can_defeat >= required:
                return True
    return False
```

Reply:
*Oh you could even sort it by length to make the early exit more likely to be sooner... ok maybe this is going too far now*

Reply:
I think I'm going to keep it as is, mostly because the current structure of `boss_requirements` is useful for debugging should there still be another error in the rule. Being able to see the end result of the validation was already very useful in debugging issues in the past, and with the speed it generates at, changing it isn't worth losing that.

Reply:
I'm saying you should have both

Reply:
Anyway think abt it, I'll do a rereview at some point tomorrow probably